File: CSharp/Text.cs
// Драйвер исходного текста
using System;
using System.IO;
using System.Text;


class Text {

   const int  TABSIZE = 3;

   public const char chSPACE = ' ';    // Пробел      
   public const char chTAB   = '\t';   // Табуляция   
   public const char chEOL   = '\n';   // Конец строки
   public const char chEOT   = '\0';   // Конец текста

   public static bool Ok = false;
   public static string Message = "Файл не открыт";
   public static int Ch = chEOT;

   private static StreamReader f;

public static void NextCh() {
   if( (Ch = f.Read()) == -1 )
      Ch = chEOT;
   else if( Ch == '\n' ) {
      Console.WriteLine();
      Location.Line++; Location.Pos = 0; Ch = chEOL;
      }
   else if( Ch == '\r' )
      NextCh();
   else if( Ch != '\t' ) {
      Console.Write((char)Ch); // !!!
      Location.Pos++;
      }
   else
      do
         Console.Write(' ');
      while( ++Location.Pos % TABSIZE != 0 );
}

public static void Reset() {
   if( Location.Path == null ) {
      Console.WriteLine(
         "Формат вызова:\n   O <входной файл>");
      Environment.Exit(1);
      }
   else if( File.Exists(Location.Path) ) {
      f = new StreamReader(Location.Path, 
      // Кириллица DOS:
         Encoding.GetEncoding(866));
      // По умолчанию:
      //    Encoding.Default);
      // Кириллица Windows:
      //    Encoding.GetEncoding(1251));
      Ok = true; Message = "Ok";
      Location.Pos = 0; Location.Line = 1;
      NextCh(); 
      }
   else {
      Ok = false;
      Message = "Входной файл не найден";
   }
}

public static void Close() {
   f.Close();
}

}

File: Java/Text.java
// Драйвер исходного текста

import java.io.*;

class Text {

   static final int  TABSIZE = 3;
   static final char chSPACE = ' ';    // Пробел      
   static final char chTAB   = '\t';   // Табуляция   
   static final char chEOL   = '\n';   // Конец строки
   static final char chEOT   = '\0';   // Конец текста

   static boolean Ok = false;
   static String Message = "Файл не открыт";
   static int Ch = chEOT;

   private static InputStream f;

   static void NextCh() {
      try {
         if( (Ch = f.read()) == -1 )
            Ch = chEOT;
         else if( Ch == '\n' ) {
            System.out.println();
            Location.Line++; Location.Pos = 0; Ch = chEOL;
            }
         else if( Ch == '\r' )
            NextCh();
         else if( Ch != '\t' ) {
            System.out.write(Ch); Location.Pos++;
            }
         else
            do
               System.out.print(' ');
            while( ++Location.Pos % TABSIZE != 0 );
      } catch (IOException e) {};
   }

   static void Reset() {
      if( Location.Path == null ) {
         System.out.println(
            "Формат вызова:\n   O <входной файл>");
         System.exit(1);
         }
      else 
         try {
            f = new FileInputStream(Location.Path);
            Ok = true; Message = "Ok";
            Location.Pos = 0; Location.Line = 1;
            NextCh(); 
         } catch (IOException e) {
            Ok = false;
            Message = "Входной файл не найден";
         }
   }

   static void Close() {
      try {
         f.close();
      } catch (IOException e) {};
   }

}
